tlscommon.c

```diff
@@ -535,13 +535,43 @@ void tlsreloadcrls() {
 	conf = (struct tls *)entry->data;
 #ifdef RADPROT_TLS
 	if (conf->tlsctx) {
+            if (conf->certkeypwd) {
+                debug(DBG_INFO, "tls: set default passwd for certkeypwd");
+                SSL_CTX_set_default_passwd_cb_userdata(conf->tlsctx, conf->certkeypwd);
+                SSL_CTX_set_default_passwd_cb(conf->tlsctx, pem_passwd_cb);
+            }
+            if (!SSL_CTX_use_certificate_chain_file(conf->tlsctx, conf->certfile) ||
+                !SSL_CTX_use_PrivateKey_file(conf->tlsctx, conf->certkeyfile, SSL_FILETYPE_PEM) ||
+                !SSL_CTX_check_private_key(conf->tlsctx)) {
+                while ((error = ERR_get_error()))
+                    debug(DBG_ERR, "tls SSL: %s", ERR_error_string(error, NULL));
+                debug(DBG_ERR, "tls: Error initialising SSL/TLS in TLS context %s", conf->name);
+                SSL_CTX_free(conf->tlsctx);
```

Comment:
This is extremely dangerous. `tlsctx` is still refered to, and the next time a connection is opened, this will get accessed; likely resulting in a segfault.

Reply:
My bad. You are right. But this is a tricky situation there. I don't know the internal details of the three APIs, SSL_CTX_use_certificate_chain_file, SSL_CTX_use_certificate_chain_file and SSL_CTX_check_private_key. Now, in the above expression, if SSL_CTX_use_certificate_chain_file() succeeds, it might change something about conf->tlsctx now assuming SSL_CTX_use_PrivateKey_file() fails for some reason, it will be too late to just bail out with a partially modified conf->tlsctx. We need some way to ensure we update all the three APIs or None.